File: PHP/skebbyAPI.php
<?php
// Skebby PHP Gateway Class
class skebbyAPI {
	
	protected $user;
	protected $pass;
	protected $http_url = "https://gateway.skebby.it/api/send/smseasy/advanced/http.php";
	
	const CURL_ERROR 	= 'cURL module is not installed. This module is necessary for the correct handling of HTTP requests by SkebbyAPI';
	const NET_ERROR 	= 'Network error: cURL can\'t connect to Skebby URL';
	const EMPTY_ERROR 	= 'One+or+more+of+the+required+variables+is+missing%3A+';
	const SENDER_ERROR 	= 'You+can+specify+only+one+type+of+sender,+numeric+or+alphanumeric';
	
	public function __construct( $username, $password ) {
		$this->user = $username;
		$this->pass = $password;
	}
	
	/**
	 * Core function where the server connects to Skebby Gateway HTTP URL
	 * and sends the data via cURL.
	 *
	 * @uses skebbyAPI::cleanPhoneNumbers( $recipients )
	 * @access protected
	 * @return string
	 */
	protected function doRequest($data) {
		try {
			$data_string = 'username='.urlencode( $this->user ).'&password='.urlencode( $this->pass );
			
			foreach( $data as $key => $val ){
				$recipients_string = '';
				if($key == 'recipients') {
					if( is_array($val[0]) && array_key_exists( 'recipient', $val[0] ) ) {
						for($i = 0; $i < count($val); $i++) {
							foreach( $val[$i] as $k => $v ) {
								$recipients_string .= '&recipients['.$i.']['.urlencode($k).']='.( ($k == 'recipient') ? $this->cleanPhoneNumbers($v) : urlencode($v) );
							}
						}
					} else {
						$recipients_string = '&recipients[]='.implode('&recipients[]=',$this->cleanPhoneNumbers($val));
					}
					$data_string .= $recipients_string;
				} else {
					$data_string .= '&'.urlencode($key).'='.urlencode($val);	
				}
			}
			
			if(!function_exists('curl_init')) { throw new Exception(self::CURL_ERROR); }
			$ch = curl_init();
			
			curl_setopt($ch,CURLOPT_CONNECTTIMEOUT,10);
			curl_setopt($ch,CURLOPT_USERAGENT,'Skebby PHP Gateway - HTTP POST method');
			curl_setopt($ch,CURLOPT_URL,$this->http_url);
			curl_setopt($ch,CURLOPT_TIMEOUT,60);
			curl_setopt($ch,CURLOPT_RETURNTRANSFER,true);
			curl_setopt($ch,CURLOPT_POST,true);
			curl_setopt($ch,CURLOPT_POSTFIELDS,$data_string);
			
			$response = curl_exec($ch);
			curl_close($ch);
			if(!$response){ throw new Exception(self::NET_ERROR); }
			return $response;
		} catch (Exception $e) {
			return $e;
		}
	}
	
	/**
	 * Main function to send the SMS. Returns a querystring or an XML, depending 
	 * on what sending method you choose before.
	 * 
	 * @uses skebbyAPI::doRequest( $data )
	 * @access public
	 * @return mixed
	 */
	public function sendSMS( $data ) {
		$error = '';
		$required = array( 
			'text', 
			'recipients'
		);
		foreach( $required as $r ) {
			if( !isset($data[ $r ]) ) {
				$error .= $r . '%2C+';
			}	
		}
		if( $error != '' ) { return (string) 'status=failed&message='.self::EMPTY_ERROR.(substr($error,0,-4)); }
		
		if(!isset($data['method'])) { $data['method'] = 'send_sms_classic'; }
		if(!is_array($data['recipients'])) { $data['recipients'] = array($data['recipients']); }
	    if($data['method'] != 'send_sms_basic') {
		    if(isset($data['sender_number']) && isset($data['sender_string'])) {
		        return 'status=failed&code=10&message='.self::SENDER_ERROR;
		    }
	    }
	    
	    return $this->doRequest($data);
	}

	/**
	 * Function to check the current balance. Returns a querystring or an XML, depending 
	 * on what sending method you choose before.
	 *
	 * @uses skebbyAPI::doRequest( $data )
	 * @access public
	 * @return string
	 */
	public function getCredit( $charset = '' ) {
		$data = array( 'method' => 'get_credit' );
		if($charset == 'UTF-8') { $data['charset'] = 'UTF-8'; }
		
	    return $this->doRequest($data);
	}
	
	/**
	 * Function to add an Alias, an alphanumeric sender_string, to be validated. Returns a querystring or an XML, depending 
	 * on what sending method you choose before.
	 *
	 * @uses skebbyAPI::doRequest( $data )
	 * @access public
	 * @return string
	 */
	public function addAlias( $data ) {
		$error = '';
		$required = array(
			'alias',
			'business_name',
			'nation',
			'vat_number',
			'taxpayer_number',
			'street',
			'city',
			'postcode',
			'contact'
		);
		foreach( $required as $r ) {
			if( !isset($data[ $r ]) ) {
				$error .= $r . '%2C+';
			}	
		}
		if( $error != '' ) { return 'status=failed&code=10&message='.self::EMPTY_ERROR.(substr($error,0,-4)); }
		
		$data['method'] = 'add_alias';
	    return $this->doRequest($data);
	}
	
	/**
	 * Function to convert the POST Request result to an associative array instead of a querystring
	 * 
	 * @access public
	 * @return array
	 */
	public function querystringToArray($data) {
		parse_str($data,$result);
		return $result;
	}
	
	/**
	 * Function to clean the phone numbers from '00' and '+' international prefixes
	 * 
	 * @access protected
	 * @return array
	 */
	protected function cleanPhoneNumbers($num) {
		for( $i = 0; $i < count($num); $i++ ) {
			$num[$i] = ltrim($num[$i], '0+');
			$num[$i] = preg_replace('/[^0-9]+/', '', $num[$i]);
		}
		return $num;
	}
}
?>

File: Java/skebbyAPI.java
// Skebby Java Gateway Class
// TO DO:
// - integrate SOAP requests
// - integrate REST requests

import java.util.*;
import java.io.*;
import java.net.*;

/**
 * skebbyAPI class 
 * it connects to the Skebby URL and uses the API to send a SMS, check the remaining credit and add an Alias (an alphanumeric sender)
 * 
 * @author Daniele Biggiogero
 * @version 1.0
 * @see URL
 * @see Map
 * @see URLEncoder
 */
public class skebbyAPI {
	// Main parameters declaration
	protected String username;
	protected String password;
	protected String charset;
	protected String url;
	
	/**
	 * skebbyAPI constructor
	 * 
	 * @param user the username necessary for authentication
	 * @param pass the password necessary for authentication
	 */
	public skebbyAPI(
		String user,
		String pass
	) {
		this(user,pass,"UTF-8");
	}
	public skebbyAPI(
		String user,
		String pass,
		String cset
	) {
		username = user;
		password = pass;
		charset = cset;
		url = "http://api.skebby.it/api/send/smseasy/advanced/http.php";
	}
	
	/**
	 * Core method.
	 * connects to skebby URL and sends data via HTTP POST request
	 * 
	 * @param data	Map<String,Object> object that contains the data
	 *				that will the sent through the request.
	 * @return 		the response from the POST request, formatted as a querystring.
	 * @throws IOException 
	 * @throws MalformedURLException if the URL is not formatted correctly
	 * @throws ProtocolException if a protocol outside <i>POST, PUT, DELETE, GET</i> is inserted
	 * @throws UnsupportedEncodingException when a charset not supported inside the URLEncoder class is used
	 */
	protected String doPostRequest( Map<String,Object> data ) throws IOException, MalformedURLException, ProtocolException, UnsupportedEncodingException {
		
        data.put("username",username);
		data.put("password",password);
        if ( !data.containsKey("charset") ) {
            data.put("charset","UTF-8");
        }
        
		if(data.containsKey("recipients")) {
			
            System.out.println(data.get("recipients").getClass().getName());
            
            if(data.get("recipients").getClass() == ArrayList.class) {
                List<Object> recipients = (ArrayList<Object>)data.get("recipients");
                System.out.println("It's an arrayList!");
                int i = 0;
                for(Object rec: recipients) {
                    System.out.println(rec);
                    i++;
                }
                
                //data.put("recipients[" + i + "]", recipients.get(i));
            } else {
                System.out.println("It's not an arrayList :(");
                String[] recipients = (String[])data.get("recipients");
            }
			/*for(int i = 0; i < recipients; i++) {
				System.out.println(i);
                //data.put("recipients[" + i + "]", recipients.get(i));
			}
            if(data.containsKey("recipients[0]") || data.containsKey("recipients[0][recipient]")) { 
                data.remove("recipients");
            }*/
		}
        
        URL connUrl = new URL(url);
		StringBuilder postData = new StringBuilder();
		for (Map.Entry<String,Object> param : data.entrySet()) {
			if (postData.length() != 0) postData.append('&');
			postData.append(URLEncoder.encode(param.getKey(), data.get("charset").toString()));
			postData.append('=');
			postData.append(URLEncoder.encode(String.valueOf(param.getValue()), data.get("charset").toString()));
		}
		byte[] postDataBytes = postData.toString().getBytes(data.get("charset").toString());

		HttpURLConnection conn = (HttpURLConnection)connUrl.openConnection();
		conn.setRequestMethod("POST");
		conn.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
		conn.setRequestProperty("User-Agent", "Skebby Java Gateway");
		conn.setRequestProperty("Content-Length", String.valueOf(postDataBytes.length));
		conn.setDoOutput(true);
		conn.getOutputStream().write(postDataBytes);

		Reader in = new BufferedReader(new InputStreamReader(conn.getInputStream(), data.get("charset").toString()));
		
		String result = new String();
		for ( int c = in.read(); c != -1; c = in.read() ) {
			result += (char)c;
		}
		return result;
	}
	
	/**
	 * sends one SMS to one or more recipients.
	 * <p>If the HTTP POST request is successful, the given querystring will return 2 cases: 
	 * <p>- Dispatch successful: status=success&remaining_sms=00
	 * <p>- Dispatch failed: status=failed&message=error+description&code=0
	 * <p>data must be included inside a Map object (HashMap, TreeMap, LinkedHashMap) to be iterated correctly
	 * 
	 * @param input the data that will be included in the doPostRequest method
	 * @return 		the result of the doPostRequest method, usually formatted as a querystring.
	 * @throws IOException 
	 * @throws IllegalArgumentException if a charset different than UTF-8 or ISO-8859-1 is used
	 */
	public String sendSMS( Map<String,Object> input ) throws IOException {
		if (input.containsKey("charset") && (!input.get("charset").equals("UTF-8") && !input.get("charset").equals("ISO-8859-1"))) {
            throw new IllegalArgumentException("Charset not supported.");
        }
        if(null == input.get("method") ) {
            input.put("method","send_sms_classic");
        }
		
		return doPostRequest( input );
	}
	
	/**
	 * check the remaining credit. 
	 * <p>If the HTTP POST request is successful, the given querystring will return:
	 * <p>status=success&credit_left=00.000&classic_sms=00&basic_sms=00&classic_plus_sms=00
	 * <p>data must be included inside a Map object (HashMap, TreeMap, LinkedHashMap) to be parsed correctly
	 * 
	 * @param input the data that will be included in the doPostRequest method
	 * @return 		the result of the doPostRequest method, usually formatted as a querystring
	 * @throws IOException 
	 * @throws IllegalArgumentException if a charset different than UTF-8 or ISO-8859-1 is used
	 */
	public String getCredit() throws IOException {
        Map<String,Object> input = new LinkedHashMap<>();
        input.put("charset","UTF-8");
		return getCredit( input );
	}
	public String getCredit( Map<String,Object> input ) throws IOException {
		if (input.containsKey("charset") && (!input.get("charset").equals("UTF-8") && !input.get("charset").equals("ISO-8859-1"))) {
            throw new IllegalArgumentException("Charset not supported.");
        }
		input.put("method", "get_credit");
		
		return doPostRequest( input );
	}
	
	/**
	 * adds an Alphanumeric sender (also known as Alias).
	 * <p>If the HTTP POST request is successful, the given querystring will return 2 cases: 
	 * <p>- Dispatch successful: status=success&alias=alias+name
	 * <p>- Dispatch failed: status=failed&message=error+description&code=0
	 * 
	 * @param input the data that will be included in the doPostRequest method
	 * @return 		the result of the doPostRequest method, usually formatted as a querystring
	 * @throws IOException 
	 * @throws IllegalArgumentException if a charset different than UTF-8 or ISO-8859-1 is used
	 */
	public String addAlias( Map<String,Object> input ) throws IOException {
		if (input.containsKey("charset") && (!input.get("charset").equals("UTF-8") && !input.get("charset").equals("ISO-8859-1"))) {
            throw new IllegalArgumentException("Charset not supported.");
        }
        input.put("method", "add_alias");
		
		return doPostRequest( input );
	}
	
	/**
	 * Prints the result in a fancy way
	 * 
	 * @param result a querystring from the doPostRequest method
	 * @return 		 prints the result in a fancier way than a querystring
	 * @throws UnsupportedEncodingException when you insert an unsupported encoding on the decode method
	 */
	 public void printResult( String result ) throws UnsupportedEncodingException {
		 String[] data = result.split("&");
		 for(String s: data) {
			String[] temp = s.split("=");
			System.out.println(temp[0] + ": " + URLDecoder.decode(temp[1],"UTF-8"));
		}
	 }
}